hostnamectl: try to set transient hostname even if setting static or pretty hostname failed

yuwata opened the pull request:
If no target (--pretty, --static, or --transient) is specified, then let's try to set transient hostname even if setting static or pretty hostname failed. This may be useful for read-only filesystem.

keszybz commented:
I like the idea of "try to set transient hostname even if updating static or pretty hostname failed", but I'm not sure if it can be done in a way that preserves the property that failure is reported. Right now, if the operation fails, there is zero feedback to the caller:
``` console
$ sudo chattr +i /etc/hostname
$ sudo hostnamectl set-hostname krowka2
$ hostnamectl --static
krowka
```

yuwata commented:
@keszybz Thank you for the comment. Now the log level is changed from `LOG_DEBUG` to `LOG_WARNING` when the failure is ignored. So, user can still get some operations are failed.

Also, I added a fix for the case when /etc/hostname or /etc/machine-info cannot be updated.

PTAL.

keszybz commented:
This is still inconsistent:
```console
$ sudo build/hostnamectl set-hostname krowka2; echo $?
Could not set static hostname, ignoring: Access denied
1
```

It  is seems like this cannot be made to work: `hostnamectl set-hostname` is certainly used a lot in scripts, and the return value matters. Right now it says "ignored" but returns error. The other option would be to always return success on failure, breaking scripts.

Maybe instead of trying to suppress the error, make the returned error more correct? Something like
> Could not set static hostname: Replacing /etc/hostname failed: Permission denied

> Could not set static hostname: Replacing /etc/hostname failed: Read-only file system
> Hint: use --transient something something.

And than the user can take appropriate action.

yuwata commented:
@keszybz Thank you for your review. Updated. Now two bus errors are introduced in hostnamed, then hostnamectl handle these error gracefully. Also, I added one commit to ignore `--transient` when `/etc/hostname` exists. PTAL.

keszybz commented on src/hostname/hostnamed.c:
```diff
@@ -367,33 +367,53 @@ static int context_update_kernel_hostname(
+                return;
+
+        **p = (struct stat) {};
+}
```
This is a good fix. I wonder if we could make it a bit less verbose... Maybe just unconditionally unset the stat cache? This means we'll re-read the file on the next query, but that's no problem.

keszybz commented on src/hostname/hostnamed.c:
```diff
@@ -788,7 +788,7 @@ static int method_set_static_hostname(sd_bus_message *m, void *userdata, sd_bus_
         r = free_and_strdup(&c->data[PROP_STATIC_HOSTNAME], name);
         if (r < 0)
-                return r;
+                return log_oom();
```
Maybe `free_and_strdup_warn` instead?

keszybz commented on src/hostname/hostnamectl.c:
```diff
@@ -291,6 +327,9 @@ static int set_hostname(int argc, char **argv, void *userdata) {
 
+        if (show_hint)
+                log_info("Hint: use --transient option when /etc/machine-info or /etc/hostname cannot be modified (e.g. located in read-only filesystem).");
+
```
Hmm, this still doesn't seem to work:
```console
$ build/hostnamectl set-hostname krowka2
Could not set static hostname: No privilege to update /etc/hostname.
$ sudo build/hostnamectl set-hostname krowka2
Could not set static hostname: No privilege to update /etc/hostname.
```
Is the hint expected in this case?

poettering commented on src/hostname/hostnamed.c:
```diff
@@ -367,33 +367,53 @@ static int context_update_kernel_hostname(
+}
+
 static int context_write_data_static_hostname(Context *c) {
+        _cleanup_(unset_statp) struct stat *s = NULL;
```
maybe add a brief comment here explaining in short words what this does. this is not obvious at all i think. i.e. maybe "make sure that if we fail here, we invalidate the cached information, since it was updated already, even if we can't make it hit the disk" or so

poettering commented on src/hostname/hostnamectl.c:
```diff
@@ -257,6 +282,21 @@ static int set_hostname(int argc, char **argv, void *userdata) {
+                if (r < 0)
+                        return r;
+
+                if (streq(source, hostname_source_to_string(HOSTNAME_STATIC))) {
```
wouldn't it be more natural to do the comparison by enum rather than by string. i.e. `hostname_source_from_string(source) == HOSTNAME_STATIC`?

yuwata commented:
@keszybz and @poettering Thank you for your review. Updated. And added one commit to use `Table`. PTAL.

poettering commented on src/libsystemd/sd-bus/bus-common-errors.c:
```diff
@@ -101,6 +101,8 @@ BUS_ERROR_MAP_ELF_REGISTER const sd_bus_error_map bus_common_errors[] = {
         SD_BUS_ERROR_MAP(BUS_ERROR_TRANSFER_IN_PROGRESS,         EBUSY),
 
         SD_BUS_ERROR_MAP(BUS_ERROR_NO_PRODUCT_UUID,              EOPNOTSUPP),
+        SD_BUS_ERROR_MAP(BUS_ERROR_NO_PRIVILEGE_TO_UPDATE_FILE,  EACCES),
```
hmm, so i was wondering: why not use SD_BUS_ERROR_ACCESS_DENIED here?

i figure because you want to communicate that it's not the client that has trouble accessing hostnamed, but hostnamed having access trouble with the file it changes. Which is probably a sensible distinction to make. But I wonder if we should maybe put even more emphasis on the file as being the culprit. i.e. BUS_ERROR_FILE_IS_PROTECTED or so? i.e. put the blame clearly on the file?

poettering commented on src/libsystemd/sd-bus/bus-common-errors.h:
```diff
@@ -81,6 +81,8 @@
 
 #define BUS_ERROR_NO_PRODUCT_UUID              "org.freedesktop.hostname1.NoProductUUID"
+#define BUS_ERROR_NO_PRIVILEGE_TO_UPDATE_FILE  "org.freedesktop.hostname1.NoPrivilegeToUpdateFile"
+#define BUS_ERROR_READONLY_FILESYSTEM          "org.freedesktop.hostname1.ReadOnlyFilesystem"
```
hmm, looking at this closer: translating CamelCase to UPPER_CASE would mean ReadOnlyFilesystem would map to READ_ONLY_FILESYSTEM.  or the otherway round: READONLY_FILESYSTEM would map to ReadonlyFilesystem. 

I'd probably stick to READ_ONLY_FILE_SYSTEM and ReadOnlyFileSystem though

poettering reviewed (commented):
looks good, but two more bikeshedding issuelets

I figure @keszybz should have a final look on this too

yuwata commented:
@poettering thank you for your review. I've renamed the two new bus errors. PTAL.

keszybz commented on src/hostname/hostnamectl.c:
```diff
@@ -280,9 +301,24 @@ static int set_hostname(int argc, char **argv, void *userdata) {
+                                if (sd_bus_error_has_names(&error,
+                                                           BUS_ERROR_FILE_IS_PROTECTED,
+                                                           BUS_ERROR_READ_ONLY_FILESYSTEM)) {
+                                        log_warning_errno(r, "Could not set static hostname, ignoring: %s", bus_error_message(&error, r));
+                                        show_hint = true;
```
I think this shouldn't be done: for most people, setting of the static hostname is the only thing that is important. The fact that the file is protected is a very good reason to fail. I think it is fine to only warn if the pretty hostname cannot be set, but not here. The hint should be shown, but the error code needs to be propagated.

@poettering, wdyt?

keszybz replied:
Actually, for consistency, I'd prefer to propagate the error in both cases. Users will mostly interact with the hostnamed api through the graphical setttings dialogues. The commandline api is for scripts and automatic provisioning and such, and there it is almost always better to simply propagate errors to the caller if there is any doubt.

yuwata commented:
@keszybz Thank you for the comments. Now the error in setting static (or pretty) hostname is propagated. PTAL.

keszybz commented:
Thanks, LGTM.

keszybz commented:
bionic-i386 failed with FAILED TESTS:  TEST-50-DISSECT.
